components/fxa-client/ios/FxAClient/FxAccountManager.swift

```diff
@@ -0,0 +1,584 @@
+    func finishAuthentication(authData: FxaAuthData, completionHandler: @escaping (Result<Void, Error>) -> Void)
+    func getAccessToken(scope: String, completionHandler: @escaping (Result<AccessTokenInfo, Error>) -> Void)
+
+    func deviceConstellation() -> DeviceConstellationProtocol?
+
+    func logout(completionHandler: @escaping (Result<Void, Error>) -> Void)
+}
+
+public extension Notification.Name {
+    static let accountLoggedOut = Notification.Name("accountLoggedOut")
+    static let accountAuthProblems = Notification.Name("accountAuthProblems")
+    static let accountAuthenticated = Notification.Name("accountAuthenticated")
+    static let accountProfileUpdate = Notification.Name("accountProfileUpdate")
+}
+
+extension Notification.Name {
+    static let accountAuthException = Notification.Name("accountAuthException")
+}
+
+protocol AccountStorage {
+    func read() -> FirefoxAccount?
+    func write(_ json: String)
+    func clear()
+}
+
+// swiftlint:disable type_body_length
+open class FxaAccountManager: FxaAccountManagerProtocol {
+    // `var` for testing.
+    var accountStorage: AccountStorage
+    let config: FxAConfig
+    let deviceConfig: DeviceConfig
+    let applicationScopes: [String]
+
+    var acct: FirefoxAccount?
+    var account: FirefoxAccount? {
+        get { return acct }
+        set {
+            acct = newValue
+            if let acc = acct {
+                constellation = makeDeviceConstellation(account: acc)
+            }
+        }
+    }
+
+    var state: AccountState = AccountState.start
+    var profile: Profile?
+    var constellation: DeviceConstellation?
+    var latestAuthState: String?
+
+    public required init(
+        config: FxAConfig,
+        deviceConfig: DeviceConfig,
+        applicationScopes: [String] = [Scope.profile, Scope.sync],
+        keychainAccessGroup: String? = nil
+    ) {
+        self.config = config
+        self.deviceConfig = deviceConfig
+        self.applicationScopes = applicationScopes
+        accountStorage = KeyChainAccountStorage(keychainAccessGroup: keychainAccessGroup)
+        setupAuthExceptionsListener()
+    }
+
+    lazy var statePersistenceCallback: FxAStatePersistenceCallback = {
+        FxAStatePersistenceCallback(manager: self)
+    }()
+
+    public func initialize(completionHandler: @escaping (Result<Void, Error>) -> Void) {
+        processEvent(event: .initialize) {
+            DispatchQueue.main.async { completionHandler(Result.success(())) }
+        }
+    }
+
+    public func hasAccount() -> Bool {
+        return state == .authenticatedWithProfile ||
+            state == .authenticatedNoProfile ||
+            state == .authenticationProblem
+    }
+
+    public func accountNeedsReauth() -> Bool {
+        return state == .authenticationProblem
+    }
+
+    public func beginAuthentication(completionHandler: @escaping (Result<URL, Error>) -> Void) {
+        DispatchQueue.global().async {
```

Comment:
If we call this in a state where it's not expected, might be worth logging that at info level for future debuging.